### PR TITLE
SAAS-3380: Added GuardDuty Export Findings Encrypted Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    category: 'GuardDuty',
+    domain: 'Management and Governance',
+    description: 'Ensure that GuardDuty Export Findings is encrypted',
+    more_info: 'GuardDuty data, such as findings, is encrypted at rest using AWS owned customer master keys (CMK). To successfully configure findings export, you must first give GuardDuty permission to use a CMK key present in your account rather than using external CMK from other accounts.',
+    link: 'https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_exportfindings.html',
+    recommended_action: 'Create GuardDuty Export Findings with customer-manager keys (CMKs) present in your account',
+    apis: ['GuardDuty:listDetectors', 'GuardDuty:listPublishingDestinations', 'GuardDuty:describePublishingDestination', 'KMS:describeKey', 'KMS:listKeys', 'STS:getCallerIdentity'],
+    settings: {
+        exported_findings_desired_encryption_level: {
+            name: 'Location GuardDuty Export Findings Desired Encryption Level',
```
Comment (sadeed12345) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:15]:
```suggestion
            name: 'GuardDuty Export Findings Desired Encryption Level',
```

File: collectors/aws/collector.js [hunk 2]
```diff
@@ -1715,14 +1715,19 @@ var postcalls = [
...
+            },
         },
         ManagedBlockchain: {
             getMember: {
                 reliesOnService: 'managedblockchain',
                 reliesOnCall: 'listNetworks',
                 override: true
             }
-        }
+        },
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:1730]:
```suggestion
        }
```

File: collectors/aws/guardduty/describePublishingDestination.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                }
+
+                collection.guardduty.describePublishingDestination[AWSConfig.region][destination.DestinationId].data = data;
+                pCb();
+            });
+        }, function(){
+            setTimeout(function(){
+                cb();
+            }, 100);
+        });
```
Comment (AkhtarAmir) [inline collectors/aws/guardduty/describePublishingDestination.js:43]:
```suggestion
cb();
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Exported Findings Encrypted',
+    category: 'GuardDuty',
+    domain: 'Management and Governance',
+    description: 'Ensure that GuardDuty Export Findings is encrypted',
+    more_info: 'GuardDuty data, such as findings, is encrypted at rest using AWS owned customer master keys (CMK). To successfully configure findings export, you must first give GuardDuty permission to use a CMK key present in your account rather than using external CMK from other accounts.',
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:9]:
```suggestion
    more_info: 'GuardDuty data, such as findings, is encrypted at rest using AWS owned customer master keys (CMK). Additionally, you can use your use key (CMKs) in order to gain more control over data encryption/decryption process.',
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    link: 'https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_exportfindings.html',
+    recommended_action: 'Create GuardDuty Export Findings with customer-manager keys (CMKs) present in your account',
+    apis: ['GuardDuty:listDetectors', 'GuardDuty:listPublishingDestinations', 'GuardDuty:describePublishingDestination', 'KMS:describeKey', 'KMS:listKeys', 'STS:getCallerIdentity'],
+    settings: {
+        exported_findings_desired_encryption_level: {
+            name: 'GuardDuty Export Findings Desired Encryption Level',
+            description: 'In order (lowest to highest) awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
+            regex: '^(awscmk|externalcmk|cloudhsm)$',
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:17]:
```suggestion
            regex: '^(awskms|awscmk|externalcmk|cloudhsm)$',
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                        continue;
+                    }
+
+                    currentEncryptionLevel = helpers.getEncryptionLevel(describeKey.data.KeyMetadata, helpers.ENCRYPTION_LEVELS);
+
+                } else {
+                    helpers.addResult(results, 3,
+                        'Unable to find encryption configuration for GuardDuty publishing destination');
+                    return cb();
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:116]:
```suggestion
                } else currentEncryptionLevel = 2; //awskms
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+                var currentEncryptionLevelString = helpers.ENCRYPTION_LEVELS[currentEncryptionLevel];
+
+                if (currentEncryptionLevel >= desiredEncryptionLevel) {
+                    helpers.addResult(results, 0,
+                        `GuardDuty Export Findings is encrypted with ${currentEncryptionLevelString} \
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:122]:
```suggestion
                        `GuardDuty findings export is configured to use ${currentEncryptionLevelString} \
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                    which is greater than or equal to the desired encryption level ${config.desiredEncryptionLevelString}`,
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        `GuardDuty Export Findings is encrypted with ${currentEncryptionLevelString} \
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:127]:
```suggestion
                        `GuardDuty findings export is configured to use ${currentEncryptionLevelString} \
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        var awsOrGov = helpers.defaultPartition(settings);
+        var accountId = helpers.addSource(cache, source, ['sts', 'getCallerIdentity', defaultRegion, 'data']);
+
+        var config = {
+            desiredEncryptionLevelString: settings.exported_findings_desired_encryption_level || this.settings.exported_findings_desired_encryption_level.default
+        };
+
+        var desiredEncryptionLevel = helpers.ENCRYPTION_LEVELS.indexOf(config.desiredEncryptionLevelString);
+        var currentEncryptionLevel;
+
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:37]: Add region loop

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+            return callback(null, results, source);
+        }
+
+        async.each(listDetectors.data, function(detectorId, cb){
+            if (!detectorId) return cb();
+ 
+            var listPublishingDestinations = helpers.addSource(cache, source,
+                ['guardduty', 'listPublishingDestinations', region, detectorId]);
+
+            if (!listPublishingDestinations || listPublishingDestinations.err || !listPublishingDestinations.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:69]:
```suggestion
            if (!listPublishingDestinations || listPublishingDestinations.err || !listPublishingDestinations.data || !listPublishingDestinations.data.Destinations) {
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                helpers.addResult(results, 3,
+                    'Unable to query for GuardDuty publishing destination lists: ' + detectorId + ': ' + helpers.addError(listPublishingDestinations), region);
+                return cb();
+            }
+
+            if (!listPublishingDestinations.data.Destinations || !listPublishingDestinations.data.Destinations.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:75]:
```suggestion
            if (!listPublishingDestinations.data.Destinations.length) {
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                `Unable to list KMS keys: ${helpers.addError(listKeys)}`, region);
+            return callback(null, results, source);
+        }
+
+        async.each(listDetectors.data, function(detectorId, cb){
+            if (!detectorId) return cb();
+ 
+            var listPublishingDestinations = helpers.addSource(cache, source,
+                ['guardduty', 'listPublishingDestinations', region, detectorId]);
+
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:68]: const resource = 'arn:' + awsOrGov + ':guardduty:' + region + ':' + accountId + ':detector/' + detectorId;

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            if (!listPublishingDestinations || listPublishingDestinations.err || !listPublishingDestinations.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query for GuardDuty publishing destination lists: ' + detectorId + ': ' + helpers.addError(listPublishingDestinations), region);
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:71]:
```suggestion
                    'Unable to query for GuardDuty publishing destination lists: ' + helpers.addError(listPublishingDestinations), region, resource);
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+                return cb();
+            }
+
+            if (!listPublishingDestinations.data.Destinations || !listPublishingDestinations.data.Destinations.length) {
+                helpers.addResult(results, 0,
+                    'you need to configure exports for s3 bucket for enabling encryption: '  + helpers.addError(listPublishingDestinations), region);
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:77]:
```suggestion
                    'Guardduty findings export is not configured'), region, resource);
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.js [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                return cb();
+            }
+
+            for (let destination of listPublishingDestinations.data.Destinations) {
+                var resource = `arn:${awsOrGov}:guardduty:${region}:${accountId}:detector/${detectorId}/publishingDestination/${ destination.DestinationId}`;
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.js:82]:
```suggestion
                let resource = `arn:${awsOrGov}:guardduty:${region}:${accountId}:detector/${detectorId}/publishingDestination/${ destination.DestinationId}`;
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/guardduty/exportedFindingsEncrypted.spec.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        it('should not return anything if list detectors response not found', function (done) {
+            const cache = createNullCache();
+            exportedFindingsEncrypted.run(cache, { exported_findings_desired_encryption_level: 'awscmk' }, (err, results) => {
+                expect(results.length).to.equal(0);
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to list KMS keys', function (done) {
+            const cache = createCache([listDetectors[0]], null, null, null, null, { message: "Unable to list KMS keys" });
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.spec.js:204]:
```suggestion
            const cache = createCache([listDetectors[0]], null, null, null, null, null, { message: "Unable to list KMS keys" });
```

File: plugins/aws/guardduty/exportedFindingsEncrypted.spec.js [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+            const cache = createCache(null, null, null, null, null, { message: 'Unable to list GuardDuty detectors'});
+            exportedFindingsEncrypted.run(cache, { exported_findings_desired_encryption_level: 'awscmk' }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                done();
+            });
+        });
+
+        it('should UNKNOWN if unable to list GuardDuty publishing destinations', function (done) {
+            const cache = createCache([listDetectors[0]], {}, describePublishingDestination[0], null, null, { message: 'Unable to query GuardDuty publishing destinations'});
```
Comment (AkhtarAmir) [inline plugins/aws/guardduty/exportedFindingsEncrypted.spec.js:187]:
```suggestion
            const cache = createCache([listDetectors[0]], {}, describePublishingDestination[0], null, null, null, null, { message: 'Unable to query GuardDuty publishing destinations'});
```

**Pull request merged**